packages/gatsby/lib/query-runner/page-query-runner.js

```diff
@@ -0,0 +1,120 @@
+/***
+ * Jobs of this module
+ * - Ensure on bootstrap that all invalid page queries are run and report
+ *   when this is done
+ * - Watch for when a page's query is invalidated and re-run it.
+ ***/
+
+const _ = require("lodash")
+const Promise = require("bluebird")
+
+const { store } = require("../redux")
+const queryRunner = require("./query-runner")
+const checkpointsPromise = require("../utils/checkpoints-promise")
+
+let queuedDirtyActions = []
+// Don't start running queries until we've reached the BOOTSTRAP_STAGE of
+// QUERY_RUNNER.
+let active = false
+const callbacks = []
+module.exports = cb => {
+  callbacks.push(cb)
+}
+
+checkpointsPromise({
+  events: [`COMPONENT_QUERIES_EXTRACTION_FINISHED`],
+}).then(() => {
+  active = true
+  const state = store.getState()
+
+  // Run queued dirty nodes now that we're active.
+  queuedDirtyActions = _.uniq(queuedDirtyActions, a => a.payload.id)
+  findAndRunQueriesForDirtyPaths(queuedDirtyActions).then(() => {
+    // Find paths without data dependencies and run them (just in case?)
+    const paths = findPathsWithoutDataDependencies()
+    // Run these pages
+    paths.forEach(path => {
+      const page = state.pages.find(p => p.path === path)
+      const component = state.pageComponents[page.component]
+      queryRunner(page, component)
+    })
+
+    // Tell everyone who cares that we're done.
+    callbacks.forEach(cb => cb())
```

Comment:
Callbacks (the array) isn't ever used again after this... which isn't obvious I agree. But this again points to the weakness of how I've been writing the (I'm not sure what's the right way to describe this) event based boot system.

I was thinking yesterday that SystemD might actually be a good model to follow since Gatsby is kind of an OS. Will read up more on that http://0pointer.de/blog/projects/systemd.html

If there's any other programming architectures you can think of for booting up an event-based system, lemme know.